### DISCUSSION TITLE
Issue with SimData including SpatialCov (raster covariate)

### DISCUSSION THREAD

**Discussion by SnowpeaSoho** [bmcclintock/momentuHMM#86] [Q&A] [answered]
Hi Brett,

I have been experimenting with momentuHMM and am very impressed with the extensive functionality you have built into this package.

At the moment, I am fitting an HMM formulation to animal tracking data including influence from a single raster covariate, with the intention to simulate from this model. The raster is spatially referenced as simple lon/lat (coord. ref. : +proj=longlat +ellps=WGS84 +datum=WGS84 +no_defs +towgs84=0,0,0). Passing the tracking dataframe to prepData() with x/y locations specified in lon/lat seems to work very sensibly. The step lengths calculated in are all sensible values in km, and the covariate values are all extracted properly from the raster with the cellFromXY call etc. The output from fitHMM() is all good.

However, when I then use this model to simulate tracks, even with very few (e.g. N=3) timesteps - which in the dataset were originally 6h intervals – immediately the error is thrown: "Movement is beyond the spatial extent of the … raster. Try expanding the extent of the raster." Further investigation with/without the SpatialCov shows that the simulation step lengths (which ought to be km) are transiting lon/lat grid coordinates as if they were easting/northing units (i.e. also specified in km). So, for example a single step length of 8km might travel from 30S, 150E straight to 157E. I guess a workaround could be to respecify the raster for simulation purposes using easting/northing, but it seems a peculiar thing to have to do at this point? So, I thought I would post it here. Issue was encountered using both CRAN version 1.4.3 (2018-11-21) and the development version 1.5.0 (2019-04-16). Apologies if I have missed something in the simData help file …

Thank you for any insight,
Sophie Bestley

**Comment by bmcclintock** [comment, marked as answer]
Hi Sophie,

Thank you for your interest in momentuHMM. Several things initially come to mind based on what you've described:

1) What are you using for the "initialPosition" argument in simData? The default is (0,0), so if your raster doesn't include (0,0) then that's going to be a problem. I'll add an error trap for when initialPosition isn't within the extent of any spatial covariates in future releases.

2) It isn't appropriate to try and use simData for simulating movement in lat/long. simData assumes Cartesian coordinates and it doesn't "know" the unit (e.g. km, m). The relative scale of movement is determined by the the data stream distribution(s) (e.g., the "mean" parameters of a gamma distribution for step length). I'll try and make this more clear in the simData help file.

3) prepData does attempt to calculate step length (in km) and turn angles when type="LL", but there's a reason the default is type="UTM" (Cartesian). The "step length" for lat/long coordinates is calculated based on Great Circle distance (using sp::spDistsN1) and the "turn angle" is based on initial bearings (using geosphere::bearing) because the direction of travel changes continuously while going along a path in lat/long.

Items 2 and 3 also apply to moveHMM. For more complicated analyses such as those supported by momentuHMM, I would generally recommend users always project their spatial data because lat/long is rather challenging to work with in animal movement models. 

Thanks for posting -- I hope this is helpful!

Cheers,
Brett

**Comment by SnowpeaSoho** [comment]
Hi Brett,

Thanks so much for the prompt response. From 1) the initial positions were being drawn from the original tracks (not 0,0) so that is fine. Thanks a lot for the clarification in 2) and 3). It is quite funny, as I had re-projected the raster etc. to try and accommodate what I thought momentuHMM would accept. But is makes sense and I will just switch to UTM throughout and see how that goes.

I have a related question for simulation purposes relating to the use of SpatialCovs. I completely understand why NAs are not allowed in covariates, in terms of log likelihoods not being able to be calculated. Hence, missing values cannot be allowed in spatial covariates. However, at the simulation stage using raster SpatialCovs, I wonder if you could foresee whether the functionality of a masking raster could be incorporated into simData. This would serve to reject a generated position at k+1 and require an alternate candidate be generated. The simData code is long, but I think this generation is happening early within the k loop where genData is used to populate X[k+1,] (~ lines 1047 – early 1100s), immediately before the spatial covariates are referenced. I can think of many ecological examples where a spatial NA exists for true reasons – e.g. no oceanic sea surface temperature would be available on land, where marine animals cannot travel. Similarly, large water bodies (lakes) should be inaccessible to models for terrestrial mammals foraging on vegetation. Or man-made barriers (e.g rabbit-proof fences) may prohibit animal movement to an immediately adjacent grid cell etc. Hence a masking raster would specify non-permitted cells for rejecting simulated moves up front, prior to the SpatialCov rasters being sampled. 

If you think this question should be diverted outside the package issues, no problem.
Thanks again,
Sophie

**Comment by bmcclintock** [comment]
An accept/reject algorithm would bias the sample from the underlying data stream distribution, so such an approach is not used in momentuHMM.  However, movement can be constrained using potential functions within a bivariate normal random walk in a similar spirit to [David Brillinger's work](https://pdfs.semanticscholar.org/66fe/74df3af9e0fb788a14778cc8f18fef239c42.pdf).  

Here's a simple example based on distance to the origin, but there's no reason the potential function couldn't account for more complicated constraints (e.g., pushing the track away from coastline, lakes, mountains, etc.). This can be implemented in momentuHMM 1.5.0 (currently only available in the 'develop') branch:

# load packages
library(momentuHMM)
library(ctmcmove)
library(plotrix)

# create raster based on distance to origin
boundary <- raster::raster(extent(-1,1,-1,1),nrows=100,ncols=100)
proj4string(boundary) <- sp::CRS("+init=epsg:4326") # required for ctmcmove::rast.grad
boundary <- raster::setValues(boundary,raster::pointDistance(sp::SpatialPoints(boundary),c(0,0),longlat=FALSE))
names(boundary) <- "boundary"
raster::plot(boundary$boundary)

# compute gradients in x- and y- directions
grad <- ctmcmove::rast.grad(boundary)
grad.x <- grad$rast.grad.x
grad.y <- grad$rast.grad.y

## specify bivariate normal biased correlated random walk with potential function based on distance to origin
dist <- list(mu="rw_mvnorm2") # bivariate normal random walk
`# mu.x_tm1 is x location at previous time step`
`# special formula function crw(mu.x_tm1,lag=1) is mu.x_tm1-mu.x_tm2 at lag 1 (previous x velocity)`
DM <- list(mu=list(mean.x=~mu.x_tm1+crw(mu.x_tm1,lag=1)+grad.x, 
                                  mean.y=~mu.y_tm1+crw(mu.y_tm1,lag=1)+grad.y, 
                                  sigma.x=~1,
                                  sigma.xy=~1,
                                  sigma.y=~1))

## specify parameters; negative coefficients for gradients result in repulsion near boundary
Par <- list(mu=c(1,0.1,-7500,1,0.1,-7500,log(0.01),0,log(0.01))) # link scale
names(Par$mu) <- c("mu.x_tm1","crw(mu.x_tm1,lag=1)","grad.x","mu.y_tm1","crw(mu.y_tm1,lag=1)","grad.y","sigma.x","sigma.xy","sigma.y")

# simulate and plot
set.seed(13,kind="Mersenne-Twister",normal.kind="Inversion")
simBound <- simData(nbStates=1, obsPerAnimal = 10000, dist=dist, Par=Par, DM=DM, spatialCovs=list(grad.x=grad.x,grad.y=grad.y), mvnCoords="mu")
plot(simBound,dataNames=c("mu.x","mu.y"),ask=FALSE)
raster::plot(boundary$boundary)
points(simBound$mu.x,simBound$mu.y,type="l")
plotrix::draw.circle(0,0,1,border="blue")

![image](https://user-images.githubusercontent.com/11667599/57952009-bb100300-78a0-11e9-8c96-1e645a0a1ed9.png)

**Comment by SnowpeaSoho** [comment]
Extreme apologies for late reply, I have been work-related out of office (hope my email reached you). Very much appreciate the time you took on the response above. Interesting point about biasing the sample. In practice I wonder whether this would be notable - perhaps more so in turn angles than step lengths? - it would be interesting to evaluate. And whether the trade-off of some bias for more biologically realistic simulation would be acceptable. 

Best practice I expect would be to include repulsion in the original HMM configuration, its quite possible that parameter estimates would be biased by NOT including constraints at this stage. Am testing multi-state HMMs, with primary interest being other environmental associations (particularly @ far sea forage grounds), so will have to think further on the extra complexity adding this in… The current bias I have implemented away from a specific (coastal water) breeding ground origin, might be swapped out for a more general coastline bias (towards/away during resident/migratory phases; for animals traversing oceans, coastlines are likely to have opposing effects per state, but moves onto land are not actually possible). Will see how I get on. Thanks again for your thoughts!

**Comment by bmcclintock** [comment]
Best of luck Sophie.  In case you're interested, I've added a more realistic coastline bias example to the vignette in momentuHMM (>= 1.5.0).
![plot_landConstraintExample](https://user-images.githubusercontent.com/11667599/59460835-d9501c80-8dd4-11e9-865d-ce148db40309.png)

**Comment by SnowpeaSoho** [comment]
Thanks Brett, I will look it up!

I meant to post actually: when I looked into UTM it turns out to be really unsuitable for animals that move beyond local/regional scales i.e. beyond one UTM zone. To supply an x-y (projected) coordinate reference sytem to momentuHMM I will therefore look to Lambert Azimuthal Equal Area (LAEA, which I am most used to using); or else perhaps Lambert Conformal Conic (LCC) which preserves distance and area properties reasonably over much larger scales. I think doing this we'll align to the assumptions of the momentuHMM code.
